server/src/main/java/org/apache/druid/server/ClientQuerySegmentWalker.java

```diff
@@ -289,17 +303,15 @@ public <T> QueryRunner<T> getQueryRunnerForSegments(final Query<T> query, final
    */
   private <T> boolean canRunQueryUsingLocalWalker(Query<T> query)
   {
-    final DataSource dataSourceFromQuery = query.getDataSource();
-    final DataSourceAnalysis analysis = dataSourceFromQuery.getAnalysis();
-    final QueryToolChest<T, Query<T>> toolChest = warehouse.getToolChest(query);
+    final DataSourceAnalysis analysis = query.getDataSourceAnalysis();
+    final QueryToolChest<T, Query<T>> toolChest = conglomerate.getToolChest(query);
 
     // 1) Must be based on a concrete datasource that is not a table.
     // 2) Must be based on globally available data (so we have a copy here on the Broker).
     // 3) If there is an outer query, it must be handleable by the query toolchest (the local walker does not handle
     //    subqueries on its own).
-    return analysis.isConcreteBased() && !analysis.isConcreteAndTableBased() && dataSourceFromQuery.isGlobal()
-           && (!(dataSourceFromQuery instanceof QueryDataSource)
-               || toolChest.canPerformSubquery(((QueryDataSource) dataSourceFromQuery).getQuery()));
+    return analysis.isConcreteBased() && !analysis.isConcreteAndTableBased() && analysis.isGlobal()
+        && toolChest.canExecuteFully(query);
```

Comment:
IIRC, the reason you need to check this `canExecuteFully` is when you have a `ScanQuery` wrapping a Union.  Generally speaking, part of the goal of introducing the `QueryLogic` is to work around the `ClientQuerySegmentWalker`'s insistence that is knows which query to run first.  Given that, I think that any query with implements `QueryLogic` should be excluded from the LocalWalker (return false) and only return true for the ClusterWalker (i.e. if the query implements `QueryLogic` and it wants to do "local" style things, the `QueryLogic` itself will implement it instead of counting on the `QuerySegmentWalker` to guess correctly).

So, I wonder if we can adjust these booleans to just check for whether the `Query` has an implementation of `QueryLogic` and if it does, then answer accordingly.  I think that will mean that you no longer need the method on `QueryToolChest` and probably also eliminate the need to implement the `QueryToolChest` at all.

Reply:
the old logic here was a bit convoluted - it was cleaned up a bit and `canExecuteFully` was part of that cleanup; it got more readable - as of now these lines are not anymore processed for `QueryLogic` cases